build: Support custom mobile Demo editor setup configuration

dcalhoun opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Support custom mobile Demo editor setup configuration.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Applying optional setup configuration files allows developers to modify
the Demo editor environment via Hooks, e.g. to set custom `initialHtml`
for the editor.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Rely upon the experimental `unstable_allowRequireContext` Metro configuration to
enable searching for the optional `setup-local.js` configuration file via
`require.context`.

This approach is a workaround as loading the optional configuration file fails
when using `allowOptionalDependencies`, as outlined in https://github.com/facebook/metro/issues/666#issuecomment-1742186952.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

**Without custom configuration**
1. Clone this branch.
1. Start the server via `npm run native start:reset`.
1. Build the app via `npm run native ios` or `npm run native android`.
1. Verify the Demo editor runs loading the existing `initial-html`
    configuration, displaying the expected blocks.

**With custom configuration**

<details><summary>Example setup configuration</summary>

```
/**
 * WordPress dependencies
 */
import { addFilter } from '@wordpress/hooks';

export default () => {
	addFilter(
		'native.block_editor_props',
		'core/react-native-editor',
		( props ) => {
			return {
				...props,
				initialHtml,
			};
		}
	);
};

const initialHtml = `
<!-- wp:heading -->
<h2 class="wp-block-heading" id="this-is-an-anchor">Just a Heading</h2>
<!-- /wp:heading -->
`;
```

</details>

1. Clone this branch.
1. Create a new file: `packages/react-native-editor/src/setup-local.js`
1. Populate the file with the "Example setup configuration" above.
1. Start the server via `npm run native start:reset`.
1. Build the app via `npm run native ios` or `npm run native android`.
1. Verify the Demo editor runs loading the local configuration file, displaying
    the blocks within the custom configuration.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
n/a

## Screenshots or screencast <!-- if applicable -->
n/a

dcalhoun committed:
build: Support custom mobile Demo editor setup configuration

Applying optional setup configuration files allows developers to modify
the Demo editor environment via Hooks, e.g. to set custom `initialHtml`
for the editor.

dcalhoun committed:
test: Disable local Demo editor configuration for tests

The `require.context` function is undefined in the test environment, as
it is a Metro-specific capability. The most straightforward way to avoid
test failures is to disable this code that relates to the local
development workflow.

fluiddot commented on packages/react-native-editor/metro.config.js:
```diff
@@ -27,6 +27,7 @@ module.exports = {
 				inlineRequires: false,
 			},
 		} ),
+		unstable_allowRequireContext: true, // Used for optional setup configuration.
```
We'd need to keep an eye on this unstable parameter when upgrading Metro (probably in the next React Native upgrade effort), in case it produces a failure in loading the local setup.

fluiddot reviewed (approved):
LGTM 🎊 ! Thanks @dcalhoun for investigating a workaround for the optional `require`, and introducing this feature 🏅 !

Apart from running the testing instructions, I also checked that modifications to the `setup-local.js` file are applied live when running the Metro server 🎊.

dcalhoun committed:
docs: Document Demo editor customization

Contributors will likely never discover or benefit from this
customization ability if it is not documented.
